### PR TITLE
NTRN-260 slinky bindings (market map & oracle)

### PR DESCRIPTION
**Pull request by quasisamurai**
Bindings for neutron. Slinky-related. [Task](https://hadronlabs.atlassian.net/browse/NTRN-260)
Oracle:

- GetPrice
- GetAllCurrencyPairs

Marketmap:

- Params
- MarketMap
- LastUpdated

Related PRs: 

- [ neutron-sdk](https://github.com/neutron-org/neutron-sdk/pull/141)
- [dev-contracts](https://github.com/neutron-org/neutron-dev-contracts/pull/47)
- [integration tests](https://github.com/neutron-org/neutron-integration-tests/pull/293)
- [njs](https://github.com/neutron-org/neutronjsplus/pull/34)

### CODE REVIEW

File: wasmbinding/bindings/query.go [hunk 1]
```diff
@@ -53,6 +59,18 @@ type QueryRegisteredQueryResultRequest struct {
...
 }
 
+type OracleQuery struct {
+	GetAllCurrencyPairs *oracletypes.GetAllCurrencyPairsRequest `json:"get_all_currency_pairs,omitempty"`
+	GetPrice            *oracletypes.GetPriceRequest            `json:"get_price,omitempty"`
+	GetPrices           *oracletypes.GetPricesRequest           `json:"get_prices,omitempty"`
+}
+
+type MarketmapQuery struct {
+	GetMarketMap   *marketmaptypes.GetMarketMapRequest   `json:"get_market_map,omitempty"`
```
Comment (NeverHappened) [inline wasmbinding/bindings/query.go]: I think it's better to rename this to `MarketMap` according to new changes:
https://github.com/skip-mev/slinky/blob/043ce53b6a9ddbab5b7ddfde8fbffebd0b9256ca/proto/slinky/marketmap/v1/query.proto#L16

File: wasmbinding/bindings/query.go [hunk 2]
```diff
@@ -53,6 +59,18 @@ type QueryRegisteredQueryResultRequest struct {
...
+	GetLastUpdated *marketmaptypes.GetLastUpdatedRequest `json:"get_last_updated,omitempty"`
```
Comment (NeverHappened) [inline wasmbinding/bindings/query.go]: I think it's better to rename this to `LastUpdated` according to new changes:
https://github.com/skip-mev/slinky/blob/043ce53b6a9ddbab5b7ddfde8fbffebd0b9256ca/proto/slinky/marketmap/v1/query.proto#L16

File: wasmbinding/queries.go [hunk 3]
```diff
@@ -206,6 +208,58 @@ func (qp *QueryPlugin) DexQuery(ctx sdk.Context, query bindings.DexQuery) (data
...
+	default:
+		return nil, wasmvmtypes.UnsupportedRequest{Kind: "unknown neutron.oracle query type"}
+	}
+}
+
+func (qp *QueryPlugin) MarketmapQuery(ctx sdk.Context, query bindings.MarketmapQuery) ([]byte, error) {
+	marketmapQueryServer := marketmapkeeper.NewQueryServer(qp.marketmapKeeper)
+
+	// handle query processing and marshaling
+	processResponse := func(resp interface{}, err error) ([]byte, error) {
```
Comment (NeverHappened) [inline wasmbinding/queries.go]: can we avoid duplication of this function? If so, it would be cleaner. But I do not insist on this

**Review by NeverHappened — CHANGES_REQUESTED** [review]

File: wasmbinding/queries.go [hunk 4]
```diff
@@ -206,6 +208,36 @@ func (qp *QueryPlugin) DexQuery(ctx sdk.Context, query bindings.DexQuery) (data
...
+
+func (qp *QueryPlugin) MarketmapQuery(ctx sdk.Context, query bindings.MarketmapQuery) ([]byte, error) {
+	marketmapQueryServer := marketmapkeeper.NewQueryServer(qp.marketmapKeeper)
+
+	switch {
+	case query.Params != nil:
+		return processResponse(marketmapQueryServer.Params(ctx, query.Params))
+	case query.LastUpdated != nil:
+		return processResponse(marketmapQueryServer.LastUpdated(ctx, query.LastUpdated))
+	case query.MarketMap != nil:
```
Comment (NeverHappened) [inline wasmbinding/queries.go:234]: there is one more query called `Market`

**Review by NeverHappened — CHANGES_REQUESTED** [review]

File: wasmbinding/queries.go (hunk 4 above)
Comment (quasisamurai) [reply wasmbinding/queries.go:234]: 🦧

**Review by NeverHappened — APPROVED** [review]

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9104146345

finally

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9104146345

**Comment by quasisamurai** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9104146345

File: wasmbinding/queries.go [hunk 5]
```diff
@@ -226,6 +260,16 @@ func dexQuery[T, R any](ctx sdk.Context, query *T, queryHandler func(ctx context
 	return data, nil
 }
 
+func processResponse(resp interface{}, err error) ([]byte, error) {
```
Comment (swelf19) [inline wasmbinding/queries.go:263]: hm, seems like if you rename `dexQuery` to something more generic like `handleQuery` you could reuse it. The functions are mostly the same.

Comment (quasisamurai) [reply wasmbinding/queries.go:263]: i personally don't want to refactor dex queries in this pr, but i like overall idea. let's create a task to refactor this?

Comment (swelf19) [reply wasmbinding/queries.go:263]: Lets create a task to refactor all bindings handlers

**Review by swelf19 — APPROVED** [review]

**Pull request merged**